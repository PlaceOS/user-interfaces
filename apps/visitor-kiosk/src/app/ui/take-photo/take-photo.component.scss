@import 'mixins';

:host {
    padding: 2em;
    display: grid;
    grid-template-columns: 1fr 2fr 1fr;
    grid-template-rows: 500px 70px;
    grid-template-areas: 'a camera b' 'c buttons d';

    video,
    canvas,
    button {
        @include fade-in('400ms');
    }

    .camera,
    .wrapper,
    .actions {
        position: relative;
        display: flex;
        align-items: center;
        justify-content: center;
    }

    .camera {
        grid-area: camera;
        flex-direction: column;
        box-shadow: 0 1px 3px 0 rgba(#000,.2), 0 1px 1px 0 rgba(#000,.14), 0 2px 1px -1px rgba(#000,.12);

        .wrapper {
            position: absolute;
            left: 2em;
            right: 2em;
            top: 4em;
            bottom: 4em;
        }
    }

    .actions {
        grid-area: buttons;
        display: flex;
        align-items: center;
        justify-content: center;
    }

    canvas,
    .confirm {
        display: none;
    }

    &.has-photo {
        video,
        .take-photo {
            display: none;
        }

        canvas,
        .confirm {
            display: unset;
        }
    }

    button {
      margin: 0 0.25em;
    }
}
