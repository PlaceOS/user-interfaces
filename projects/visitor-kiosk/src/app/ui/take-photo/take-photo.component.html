<div class="camera">
  <mat-spinner diameter="32"></mat-spinner>
  <div class="text">Please wait...</div>
  <div class="wrapper">
    <video id="video" #video autoplay width=200 height=355></video>
    <canvas id="canvas" #canvas width=200 height=355></canvas>
  </div>
</div>
<div class="actions">
  <button class="take-photo" mat-button (click)="takePhoto()">Take Photo</button>
  <button class="confirm" mat-button (click)="cancelPhoto()">Cancel</button>
  <button class="confirm" mat-button (click)="acceptPhoto()">Accept</button>
</div>