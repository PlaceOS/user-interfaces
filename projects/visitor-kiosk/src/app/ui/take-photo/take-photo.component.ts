import {
    Component,
    OnInit,
    ViewChild,
    HostBinding,
    OnDestroy,
    Input,
    Output,
    EventEmitter
} from '@angular/core';

@Component({
    selector: 'a-take-photo',
    templateUrl: './take-photo.component.html',
    styleUrls: ['./take-photo.component.scss']
})
export class TakePhotoComponent implements OnInit, OnDestroy {
    @Output() public photoAccepted = new EventEmitter();
    @HostBinding('class.has-photo') public hasPhoto = false;
    public video: any;
    public canvas: any;
    private canvasContext = null;

    private constraints = {
        audio: false,
        video: {
            width: { min: 200, max: 200 },
            height: { min: 355, max: 355 }
        }
    };

    public imgSrc = null;

    constructor() {}

    public ngOnInit() {
        setTimeout(() => this.getCanvasContext(), 500);
    }

    public ngOnDestroy() {
        this.stopCapture();
        this.video = null;
        this.canvasContext = null;
        this.canvas = null;
        this.imgSrc = null;
    }

    private startCapture() {
        this.imgSrc = null;
        console.log('TakePhotoComponent::startCapture');
        this.video = document.getElementById('video') as any;
        if (!this.video) {
            return setTimeout(() => this.startCapture(), 200);
        } else {
            navigator.mediaDevices.getUserMedia(this.constraints).then(stream => {
                this.video.srcObject = stream;
            });
        }
    }

    private stopCapture() {
        console.log('TakePhotoComponent::stopCapture');
        if (this.video && this.video.srcObject) {
            this.video.srcObject.getVideoTracks().forEach(track => track.stop());
        }
    }

    private getCanvasContext() {
        console.log('TakePhotoComponent::getCanvasContext');
        this.canvas = document.getElementById('canvas') as any;
        if (!this.canvas) {
            return setTimeout(() => this.getCanvasContext(), 200);
        }
        this.canvasContext = this.canvas.getContext('2d');
        setTimeout(() => this.startCapture(), 200);
    }

    public takePhoto() {
        console.log('TakePhotoComponent::takePhoto');
        this.canvasContext.drawImage(this.video, 0, 0, this.canvas.width, this.canvas.height);
        this.hasPhoto = true;
        this.stopCapture();
    }

    public cancelPhoto() {
        console.log('TakePhotoComponent::cancelPhoto');
        this.canvasContext.clearRect(0, 0, this.canvas.width, this.canvas.height);
        this.hasPhoto = false;
        this.startCapture();
    }

    public acceptPhoto() {
        try {
            this.imgSrc = this.canvas.toDataURL('image/jpeg', 0.75);
            this.photoAccepted.emit(this.imgSrc);
            console.log('TakePhotoComponent::acceptPhoto Success');
        } catch (err) {
            console.error('TakePhotoComponent::acceptPhoto Error converting image', err);
            this.photoAccepted.emit(null);
            this.cancelPhoto();
        }
    }
}
