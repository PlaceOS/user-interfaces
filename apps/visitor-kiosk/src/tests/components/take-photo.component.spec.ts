import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { TakePhotoComponent } from '../../app/components/take-photo.component';

describe('TakePhotoComponent', () => {
    let spectator: Spectator<TakePhotoComponent>;
    const createComponent = createComponentFactory({
        component: TakePhotoComponent,
        imports: [MatProgressSpinnerModule]
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
    });
});
