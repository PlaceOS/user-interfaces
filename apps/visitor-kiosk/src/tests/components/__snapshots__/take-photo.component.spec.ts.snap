// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TakePhotoComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="flex flex-col items-center justify-center rounded overflow-hidden relative"
    name="camera"
  >
    <mat-spinner
      class="mat-spinner mat-progress-spinner mat-primary mat-progress-spinner-indeterminate-animation _mat-animation-noopable"
      diameter="32"
      mode="indeterminate"
      role="progressbar"
      style="width: 32px; height: 32px;"
    >
      <svg
        aria-hidden="true"
        focusable="false"
        preserveAspectRatio="xMidYMid meet"
        style="width: 32px; height: 32px;"
        viewBox="0 0 25.2 25.2"
      >
        <circle
          cx="50%"
          cy="50%"
          r="11"
          style="animation-name: mat-progress-spinner-stroke-rotate-32; stroke-dasharray: 69.11503837897544px; stroke-width: 10%;"
        />
        
        
      </svg>
    </mat-spinner>
    <div
      class="text"
    >
      Please wait...
    </div>
    <video
      autoplay=""
      class="absolute inset-0"
      height="400"
      id="video"
      width="360"
    />
    <canvas
      class="absolute inset-0"
      height="400"
      id="canvas"
      width="360"
    />
  </div>
  <div
    class="flex items-center justify-center mt-4 space-x-2"
  >
    <button
      class="take-photo"
      mat-button=""
    >
       Take Photo 
    </button>
    
  </div>
  
</div>
`;
